tests/cpp/tree/test_constraints.cu

```diff
@@ -0,0 +1,321 @@
+    ASSERT_EQ(h_result[1], 1);
+    ASSERT_EQ(h_result[2], 2);
+  }
+
+  {
+    tree::TrainParam large_param = GetParameter();
+    large_param.interaction_constraints = R"([[1, 139], [244, 0], [139, 221]])";
+    FConstraintWrapper large_features(large_param, 256);
+    large_features.Split(0, 139, 1, 2);
+    auto span = large_features.QueryNode(0);
+    std::vector<int32_t> h_result (span.size());
+    thrust::copy(thrust::device_ptr<int32_t>(span.data()),
+                 thrust::device_ptr<int32_t>(span.data() + span.size()),
+                 h_result.begin());
+    ASSERT_EQ(h_result.size(), 3);
+    ASSERT_EQ(h_result[0], 1);
+    ASSERT_EQ(h_result[1], 139);
+    ASSERT_EQ(h_result[2], 221);
+  }
+}
+
+namespace {
+
+void CompareFeatureList(common::Span<int32_t> s_output, std::vector<int32_t> solution) {
+  std::vector<int32_t> h_output(s_output.size());
+  thrust::copy(thrust::device_ptr<int32_t>(s_output.data()),
+               thrust::device_ptr<int32_t>(s_output.data() + s_output.size()),
+               h_output.begin());
+  ASSERT_EQ(h_output.size(), solution.size());
+  for (size_t i = 0; i < solution.size(); ++i) {
+    ASSERT_EQ(h_output[i], solution[i]);
+  }
+}
+
+}  // anonymous namespace
+
+TEST(FeatureInteractionConstraint, Query) {
+  {
+    tree::TrainParam param = GetParameter();
+    int32_t constexpr kFeatures = 6;
+    FConstraintWrapper constraints(param, kFeatures);
+    std::vector<int32_t> h_input_feature_list {0, 1, 2, 3, 4, 5};
+    dh::device_vector<int32_t> d_input_feature_list (h_input_feature_list);
+    common::Span<int32_t> s_input_feature_list = dh::ToSpan(d_input_feature_list);
+
+    auto s_output = constraints.Query(s_input_feature_list, 0);
+    CompareFeatureList(s_output, h_input_feature_list);
+  }
+  {
+    tree::TrainParam param = GetParameter();
+    int32_t constexpr kFeatures = 6;
+    FConstraintWrapper constraints(param, kFeatures);
+    constraints.Split(/*node_id=*/0, /*feature_id=*/1, /*left_id=*/1, /*right_id=*/2);
+    constraints.Split(/*node_id=*/1, /*feature_id=*/0, /*left_id=*/3, /*right_id=*/4);
+    constraints.Split(/*node_id=*/4, /*feature_id=*/3, /*left_id=*/5, /*right_id=*/6);
+    /*
+     * (node id) [allowed features]
+     *
+     *               (0) [1, 2]
+     *           /        \
+     *      {split at 0}   \
+     *         /            \
+     *        (1)[0, 1, 2]  (2)[1, 2]
+     *    /        \
+     *   /      {split at 3}
+     *  /            \
+     * (3)[0, 1, 2]  (4)[0, 1, 2, 3, 4, 5]
+     *
+     */
+
+    std::vector<int32_t> h_input_feature_list {0, 1, 2, 3, 4, 5};
+    dh::device_vector<int32_t> d_input_feature_list (h_input_feature_list);
+    common::Span<int32_t> s_input_feature_list = dh::ToSpan(d_input_feature_list);
+
+    auto s_output = constraints.Query(s_input_feature_list, 1);
+    CompareFeatureList(s_output, {0, 1, 2});
+    s_output = constraints.Query(s_input_feature_list, 2);
+    CompareFeatureList(s_output, {1, 2});
```

Comment:
@trivialfis this is the beginning of the failure.

note: if i remove line 250, then this check will fail if i query on node 1 as well